src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSpecification.java

```diff
@@ -0,0 +1,161 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.cft.query;
+
+import org.springframework.data.jpa.domain.Specification;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.access.entities.AccessControlResponse;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.permission.entities.PermissionTypes;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.role.entities.RoleAssignment;
+import uk.gov.hmcts.reform.wataskmanagementapi.auth.role.entities.enums.Classification;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.entities.TaskResource;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.entities.TaskRoleResource;
+import uk.gov.hmcts.reform.wataskmanagementapi.cft.enums.CFTTaskState;
+import uk.gov.hmcts.reform.wataskmanagementapi.controllers.request.SearchTaskRequest;
+import uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.camunda.SecurityClassification;
+import uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search.SearchParameter;
+import uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search.SearchParameterKey;
+
+import java.util.ArrayList;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+import javax.persistence.criteria.Join;
+import javax.persistence.criteria.Predicate;
+
+@SuppressWarnings({"PMD.DataflowAnomalyAnalysis"})
+public final class TaskResourceSpecification {
+
+    private TaskResourceSpecification(){
+        // avoid creating object
+    }
+
+    public static Specification<TaskResource> getTasks(
+            SearchTaskRequest searchTaskRequest,
+            AccessControlResponse accessControlResponse,
+            List<PermissionTypes> permissionsRequired
+    ) {
+
+        return searchByRoles(permissionsRequired, accessControlResponse)
+            .and(searchByClassification(accessControlResponse)
+            .and(searchByRegion()
+            .and(searchByState(searchTaskRequest)
+            .and(searchByLocation(searchTaskRequest)
+            .and(searchByJurisdiction(searchTaskRequest)
+            .and(searchByCaseId(searchTaskRequest)
+            .and(searchByUser(searchTaskRequest))))))));
```

Comment:
Maybe I don't understand exactly how this works but just wanted to understand.  If we have multiple role assignments in the accessControlResponse then I guess we want to `OR` at the top level?  Is that what we are doing here?

Reply:
RoleAssignments condition can be placed anywhere in where clause and satisfy with multiple role assignment OR condition, for eg: based on above filters the hibernates generates query 

    where
        (
            taskresour0_.assignee in (
                ?
            )
        ) 
        and 1=1 
        and (
            taskresour0_.jurisdiction in (
                ?
            )
        ) 
        and (
            taskresour0_.location in (
                ?
            )
        ) 
        and (
            taskresour0_.state in (
                ?
            )
        ) 
        and taskresour0_.region=? 
        and (
            taskresour0_.security_classification in (
                ? , ? , ?
            )
        ) 
        **and (
            taskrolere1_.role_name=? 
            or taskrolere1_.role_name=?
        )** 
        and taskrolere1_.read=true 
    order by
        taskresour0_.task_id asc limit ?

Reply:
check this condition in the above query 
        and (
            taskrolere1_.role_name=? 
            or taskrolere1_.role_name=?
        )